common/js/src/messaging/single-message-based-channel-unittest.js

```diff
@@ -0,0 +1,310 @@
+/** @license
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+goog.require('GoogleSmartCard.SingleMessageBasedChannel');
+goog.require('GoogleSmartCard.MessageChannelPinging.Pinger');
+goog.require('GoogleSmartCard.MessageChannelPinging.PingResponder');
+goog.require('GoogleSmartCard.TypedMessage');
+goog.require('goog.Promise');
+goog.require('goog.async.Delay');
+goog.require('goog.testing.PropertyReplacer');
+goog.require('goog.testing.jsunit');
+goog.require('goog.testing.mockmatchers');
+
+goog.setTestOnly();
+
+goog.scope(function() {
+
+/** @const */
+var GSC = GoogleSmartCard;
+
+/** @const */
+var SingleMessageBasedChannel = GSC.SingleMessageBasedChannel;
+
+/** @const */
+var Pinger = GSC.MessageChannelPinging.Pinger;
+
+/** @const */
+var PingResponder = GSC.MessageChannelPinging.PingResponder;
+
+/** @const */
+var TypedMessage = GSC.TypedMessage;
+
+/** @const */
+var EXTENSION_ID = 'DUMMY_EXTENSION_ID';
+
+/** @const */
+var TEST_MESSAGES = [
+    new TypedMessage('service x', {}),
+    new TypedMessage('service x', {some_data: 1}),
+    new TypedMessage('service y', {})];
+
+/** @const */
+var channelId = PingResponder.generateChannelId();
+
+/** @const */
+var pingResponseMessage = new TypedMessage(
+    PingResponder.SERVICE_NAME,
+    PingResponder.createMessageData(channelId)).makeMessage();
+
+/**
+ * Mock matcher that matches strings.
+ * @const
+ */
+var isStringMockMatcher = new goog.testing.mockmatchers.ArgumentMatcher(
+    function(value) {
+      return typeof value === 'string';
+    }, 'isStringMessage');
+
+/**
+ * Mock matcher that matches only the ping messages.
+ * @const
+ */
+var isPingMessageMockMatcher = new goog.testing.mockmatchers.ArgumentMatcher(
+    function(value) {
+      var typedMessage = TypedMessage.parseTypedMessage(value);
+      return typedMessage && typedMessage.type == Pinger.SERVICE_NAME;
+    }, 'isPingMessage');
+
+var globalChannel;
+
+/**
+ * Setup the mock for chrome.runtime.sendMessage.
+ * @param {!goog.promise.Resolver} testCasePromiseResolver
+ * @return {goog.testing.PropertyReplacer}
+ */
+function setupSendMessageMock(testCasePromiseResolver) {
+  var propertyReplacer = new goog.testing.PropertyReplacer;
```

Comment:
I'd prefer having the property replacer constructed at the test level, not in this helper function.
First, because this works well only for a single setup function; second, because it's clearer when the setup and the teardown steps can be seen from the actual test case code.